app/controllers/courses_controller.rb

```diff
@@ -5,25 +5,51 @@
 
 class CoursesController < ApplicationController
   # you need to be able to pick a course to be authorized for it
-  skip_before_action :authorize_user_for_course, only: [:index, :new, :create]
+  skip_before_action :authorize_user_for_course, only: [:index, :show, :new, :create, :join]
   # if there's no course, there are no persistent announcements for that course
-  skip_before_action :update_persistent_announcements, only: [:index, :new, :create]
+  skip_before_action :update_persistent_announcements, only: [:index, :new, :create, :join]
 
   def index
     courses_for_user = User.courses_for_user current_user
 
-    if courses_for_user.any?
+    if not courses_for_user.nil?
       @listing = categorize_courses_for_listing courses_for_user
     else
-      redirect_to(home_no_user_path) && return
+      @listing = []
     end
 
-    render layout: "home"
+    # render layout: "home"
   end
 
-  action_auth_level :show, :student
+  action_no_auth :show
   def show
-    redirect_to course_assessments_url(@course)
+
+    @course = Course.find_by(name: params[:name])
+
+    unless @course
```

Comment:
Use the modifier form here:
`redirect_to(controller: :home, action: :error) && return unless @course`